synapse/handlers/user_directory.py

```diff
@@ -392,6 +392,7 @@ async def _track_user_joined_room(self, room_id: str, user_id: str) -> None:
         if is_public:
             await self.store.add_users_in_public_rooms(room_id, (user_id,))
         else:
+            # TODO: get_local_users_in_room here
```

Comment:
TODO: Split this out to another PR

Reply:
The logic is a bit confusing and subtle here but we do actually use remote users here,

https://github.com/matrix-org/synapse/blob/6f0c3e669da458e838e7b4b165a13e8a5312d6d0/synapse/handlers/user_directory.py#L407-L410

Refactored the code a bit here to make this fact more clear: https://github.com/matrix-org/synapse/pull/13966